internal/extender/resource.go

```diff
@@ -336,92 +337,66 @@ func (s *SparkSchedulerExtender) potentialNodes(availableNodesSchedulingMetadata
-		}
-		driver, err := s.podLister.getDriverPod(ctx, executor)
-		if err != nil {
-			return "", failureInternal, err
+	if alreadyBoundNode != "" {
+		// check that it is part of the nodes passed and return it
+		if resultNode, ok := s.getReservationNodeFromNodeList([]string{alreadyBoundNode}, nodeNames); ok {
+			return resultNode, successAlreadyBound, nil
 		}
-		sparkResources, err := sparkResources(ctx, driver)
-		if err != nil {
-			return "", failureInternal, err
-		}
-		if outcome == failureUnbound && (sparkResources.minExecutorCount+extraExecutorCount) < sparkResources.maxExecutorCount {
-			// dynamic allocation case where driver is requesting more executors than min but less than max
-			node, outcome, err := s.rescheduleExecutor(ctx, executor, nodeNames, sparkResources, false)
-			if err != nil {
-				if outcome == failureFit {
-					return node, failureFitExtraExecutor, werror.Error("not enough capacity to schedule the extra executor")
-				}
-				return node, outcome, err
-			}
-			softReservation := v1beta1.Reservation{
-				Node:   node,
-				CPU:    sparkResources.executorResources.CPU,
-				Memory: sparkResources.executorResources.Memory,
-			}
-			err = s.softReservationStore.AddReservationForPod(ctx, driver.Labels[common.SparkAppIDLabel], executor.Name, softReservation)
+	}
+
+	unboundReservationNodes, err := s.resourceReservationManager.FindUnboundReservationNodes(ctx, executor)
+	if err != nil {
+		return "", failureInternal, werror.WrapWithContextParams(ctx, err, "error when looking for unbound reservations")
+	}
+	if len(unboundReservationNodes) > 0 {
+		if resultNode, ok := s.getReservationNodeFromNodeList(unboundReservationNodes, nodeNames); ok {
+			err := s.resourceReservationManager.ReserveForExecutor(ctx, executor, resultNode)
 			if err != nil {
-				return "", failureInternal, err
+				return "", failureInternal, werror.WrapWithContextParams(ctx, err, "failed to reserve node for executor")
 			}
-			// We might have created a demand object for this executor when we were under min count, so we should remove if it exists
 			s.removeDemandIfExists(ctx, executor)
-			return node, successScheduledExtraExecutor, nil
+			return resultNode, success, nil
 		}
-		return "", outcome, unboundResErr
-	}
-	// the reservation to be selected for the current executor needs to be on a node that the extender has received from kube-scheduler
-	nodeToReservation := make(map[string]string, len(unboundReservations))
-	for _, reservationName := range unboundReservations {
-		nodeToReservation[resourceReservation.Spec.Reservations[reservationName].Node] = reservationName
 	}
-	var unboundReservation string
-	for _, name := range nodeNames {
-		if reservation, ok := nodeToReservation[name]; ok {
-			unboundReservation = reservation
-			break
-		}
+
+	// Else, check if you still can have an executor, and if yes, reschedule
+	freeExecutorSpots, err := s.resourceReservationManager.GetFreeExecutorSpots(ctx, executor)
```

Comment:
What's the difference between `freeExecutorSpots` and `unboundReservations`?

Reply:
`unboundReservationNodes` is a set of nodes that are reserved but that don't have executors currently scheduled on them.
`GetFreeExecutorSpots` returns the number of remaining executors this Spark application is allowed to schedule. Agreed that the naming is confusing, renamed to `GetRemainingAllowedExecutorCount`.